### PR TITLE
fix #6164: Reduce format failure for non default `imports_granularity`

### LINKED ISSUE #6164: `imports_granularity = "Module"` doesn't work if module path is too long
## Repro steps

1. Add the following inputs to `tests/source/5131_module.rs`.
2. Execute `cargo test test::system_tests` and get output.

### case 1. correct

Input:

```
use foo::{
    Foo,
    bar::Bar,
};
```

Output:

```
use foo::bar::Bar;
use foo::Foo;
```

### case 2. wrong

Input:

```
use foooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::{
    Foo,
    bar::Bar,
};
```

Output:

```
use foooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::{
    Foo,
    bar::Bar,
};
```

Expected:

```
use foooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::bar::Bar;
use foooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::Foo;
```

### case 3. wrong

Input:

```
use x::fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::{
    Foo,
    bar::Bar,
};
```

Output:

```
use x::fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::{
    Foo,
    bar::Bar,
};
```

Expected:

```
use x::fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::bar::Bar;
use x::fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo::Foo;
```

Link: #4991

### PR DESCRIPTION
**Pull request by kenoss**
Fixed: #6164

This patch reduces format failure for non default `imports_granularity` and correct the behavior around some edge cases.

- Supports too long line of `use ...` that contains `{}`.
- Fixes the width calculation of too long lines of `use ...`.

### CODE REVIEW

File: tests/target/issue_3033.rs [hunk 1]
```diff
@@ -1,2 +1,12 @@
-use dom::bindings::codegen::Bindings::BluetoothRemoteGATTServerBinding::BluetoothRemoteGATTServerBinding::
-    BluetoothRemoteGATTServerMethods;
+use dom::bindings::codegen::Bindings::BluetoothRemoteGATTServerBinding::
+    BluetoothRemoteGATTServerBinding::BluetoothRemoteGATTServerMethods;
```
Comment (kenoss) [inline tests/target/issue_3033.rs:2]: Exceeds 100 chars. It was preserved because the format failed.

Comment (ytmimi) [reply tests/target/issue_3033.rs:2]: Given that this changes existing behavior these changes need to be [version gated](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#version-gate-formatting-changes)

**Review by ytmimi — CHANGES_REQUESTED** [review]
I appreciate you jumping in to work on this, though I'm hoping we can find a way to  refactor and simplify the implementation.

I also want to be mindful of your time and say up front that it might be a while before I'm able to get around to a follow up review on this.

File: src/shape.rs [hunk 2]
```diff
@@ -242,6 +242,13 @@ impl Shape {
         })
     }
 
+    pub(crate) fn add_width(&self, width: usize) -> Shape {
+        Shape {
+            width: self.width + width,
+            ..*self
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kenoss) [reply src/shape.rs:250]: Consider this case:

```
// With max width                     40
//                                     |
use long_segment_loooooooooooooooooong::{Foo, Bar};

// should be formated to
use long_segment_loooooooooooooooooong::{
    Foo,
    Bar,
};
```

Look at the `UseTree::rewrite()`, especially

```
let s = if prev_is_allow_overflow
    && matches!(segment.kind, UseSegmentKind::List(_))
{
    shape.add_width(1)
} else {
    shape.clone()
};
```

Let's trace the behavior of the format above.

```
committed = "use "

// 35 = 40 - (4 + 1)
shape = {indent: 0, offset: 4, width: 35}
// len = 34
segment = "long_segment_loooooooooooooooooong"

// 34 + 2 exceeds 35, but the above segment should be stacked to the current line because this is the first segment
committed = "use long_segment_loooooooooooooooooong::"
// with overflow.
shape = {indent: 0, offset: 40, width: 0}
prev_is_allow_overflow = true

// TIMING1
segment = "{Foo, Bar}"
// Here, we must secure the room for "{" since `prev_is_allow_overflow == true`, or `segment.rewrite(context, shape)?` fails.
s = {indent: 0, offset: 40, width: 1}

committed = "use long_segment_loooooooooooooooooong::{\n    Foo,\n    Bar,\n}"
```

Alternative options:

1. Use `segment.rewrite(context, shape.infinite_width())?;` at the `TIMING1` instead.

We can't do that because it produces

```
use long_segment_loooooooooooooooooong::{Foo, Bar};
```

2. Add a variant of `UseSegment::rewirte()` that can handle yet another argument `should_commit_open_brace_even_if_no_width`.

It's not simple compared to adding `Shape::add_width()`...

So, I added `Shape::add_width()`.

File: src/shape.rs [hunk 3]
```diff
@@ -254,6 +261,10 @@ impl Shape {
         self.add_offset(width).sub_width(width)
     }
 
+    pub(crate) fn offset_left_maybe_overflow(&self, width: usize) -> Shape {
+        self.add_offset(width).saturating_sub_width(width)
+    }
```
[the remark this answers is not in the document]
Comment (kenoss) [reply src/shape.rs:266]: Similar. There are the cases that we need to allow overflow. We can proceed `Shape` with this method without fail.

File: src/imports.rs [hunk 4]
```diff
@@ -1077,15 +1077,75 @@ impl Rewrite for UseSegment {
...
+                        segment_str
+                    );
+                    if iter.peek().is_some() {
+                        chunk.push_str("::");
+                    }
+                    result.push_str("\n");
+                    result.push_str(&chunk);
+                    shape = shape_top_level.offset_left_maybe_overflow(segment_str.len());
+                    prev_is_allow_overflow = true;
+                }
```
[the remark this answers is not in the document]
Comment (kenoss) [reply src/imports.rs:1148]: Let me think...

**Comment by kenoss** [comment]
Thank you for quick look!

I'll rethink and refactor. I'll mention once I've done.

File: src/imports.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kenoss) [reply src/imports.rs]: Refactored. Looking ahead next + retry.

**Comment by kenoss** [comment]
@ytmimi Could you have a look? Thanks!

**Comment by ytmimi** [comment]
> @ytmimi Could you have a look? Thanks!

I really appreciate you taking the time to make these changes. As I mentioned in https://github.com/rust-lang/rustfmt/pull/6165#pullrequestreview-2067260236 I likely won't have time to re-review this for a little while, but I'll follow up when I can

**Comment by kenoss** [comment]
np. Take your time.

**Comment by kenoss** [comment]
Rebased and conflict resolved. Updated using `RewriteResult/RewriteError`.

**Comment by kenoss** [comment]
This is a friendly ping.

Three months passed.

@ytmimi Could you have a look? Let me know if I can help.

**Comment by kenoss** [comment]
I understand your situation. (But... I wonder why don't you add more reviewers? It looks other PRs also are stuck with review.)

OK. Then, I'll pause to make this PR fresh. Let me know when you can review. Then, I'll rebase it.

File: tests/target/issue_3033.rs (hunk 1 above)
Comment (kenoss) [reply tests/target/issue_3033.rs]: Ah, sorry. I missed the comment. Fixed.

**Comment by ytmimi** [comment]
@kenoss Thanks for taking the time to rebase the PR. At the moment this is a lower priority item for the team and I don't anticipate I'll get around to reviewing the implementation before the end of the year. It's hard to say when exactly I'll be able to revisit this one, but I'll be sure to reach out when I do.

**Comment by kenoss** [comment]
Thank you for taking time!

FYI, it's a blocker for me to upstream it because I need it for CI. I don't want to diverge it.

File: src/imports.rs [hunk 5]
```diff
@@ -1126,18 +1126,107 @@ impl Rewrite for UseTree {
...
+                next_segment: Option<&&UseSegment>,
+            ) -> Result<(String, Shape), RewriteError> {
+                let mut rewritten_segment = curr_segment.rewrite_result(context, shape.clone())?;
+                if next_segment.is_some() {
+                    rewritten_segment.push_str("::");
+                }
+                let reserved_room_for_brace = match next_segment.map(|s| &s.kind) {
+                    Some(UseSegmentKind::List(_)) => "{".len(),
+                    _ => 0,
+                };
```
Comment (chansuke) [inline src/imports.rs:1145]: It would be helpful to add a comment explaining why `reserved_room_for_brace` is calculated?

**Review by chansuke — APPROVED** [review]
I’ve reviewed your changes, including the test cases, and overall, the changes look good to me.

File: src/imports.rs (hunk 5 above)
Comment (kenoss) [reply src/imports.rs:1147]: Added. Thanks!

**Comment by kenoss** [comment]
Thanks! Comment added. All review comments resolved. (I'm not familiar with GitHub and a way of this repository. Which should I or you do `Resolve conversation`? I'll leave them.)

**Comment by chansuke** [comment]
Yes, you can `Resolve conversation` the outdated conversation

**Comment by kenoss** [comment]
Rebased to current HEAD (fd0ea74) for projects using 2024 edition.
